### DISCUSSION TITLE
Support Files / Images in Repeater Interface

### DISCUSSION THREAD

**Discussion by flayks** [directus/directus#3001] [Feature Requests]
I don't really know if it's planned or not at this stage of the v9, but I saw somewhere on the main v8 repo that files and images would be added to the (JSON) Repeater field? That would be extremely helpful as often clients want to edit versatile pieces of informations, sometimes including an image or images. Having used Prismic for a recent project, the ability to be very flexible with this kind of data is gold.

**Comment by benhaynes** [comment]
I may be wrong, but I think that all relational data and fields will be reserved for the _relational_ repeater interface, not JSON. @rijkvanzanten ?

**Reply by rijkvanzanten** [reply]
Correct! You should think of the regular repeater as a nice interface around a "text blob". The "fields" inside the regular repeater aren't actual API fields (eg columns), and therefore don't support any of Directus' API features like sorting/filter/relations etc

**Reply by flayks** [reply]
Right. That makes sense, thanks. Is that already implemented or is it in the works?

**Reply by stouch** [reply]
@rijkvanzanten  You mean that it will never be possible to import a file UUID (from a library selector) into a json-repeater uuid field ? We will have to change (remove/ recreate) the type of the field , with a relational repeater ? :'(

**Reply by Kematia** [reply]
> @rijkvanzanten You mean that it will never be possible to import a file UUID (from a library selector) into a json-repeater uuid field ? We will have to change (remove/ recreate) the type of the field , with a relational repeater ? :'(

Yup, you can do it with an m2a interface right now!

**Reply by benhaynes** [reply]
> @rijkvanzanten You mean that it will never be possible to import a file UUID (from a library selector) into a json-repeater uuid field ? We will have to change (remove/ recreate) the type of the field , with a relational repeater ? :'(

You can absolutely do this in Directus, it just won't be in the core code base because it doesn't pass our 80/20 rule. To do this, just duplicate their current JSON repeater interface and adjust to support files... This is exactly what extensions are for.

**Reply by stouch** [reply]
@benhaynes  Ok , so, you mean, I just need to create a extension from the repeater original code ( https://github.com/directus/directus/blob/main/app/src/interfaces/list/list.vue ) and to modify `fieldsWithNames` to handle uuid with a file selector ? Am I wrong ? Do you know how I could do it more precisely ? Thank you

**Reply by benhaynes** [reply]
I'm not a proper developer, but yes, that's the idea. It doesn't make sense for us to support that in core, but nothing stopping you from doing it in an extension.

**Reply by stouch** [reply]
Do you know if someone has ever done something like this ? It seems easy to develop it, and very useful for people who want a simple integration of JSON repeater (without creating a M2A and mysql tables etc), just an interface of File Selector in uuid Repeater field. I installed directus through docker, and I do not have much time this week but I 'll try to develop this anyway if nothing exists.

<img width="622" alt="Capture d’écran 2021-11-16 à 16 47 14" src="https://user-images.githubusercontent.com/17531455/142018541-1d9f9b2d-40e5-479d-8170-799e22799dce.png">
.

**Reply by emilianovelasco** [reply]
Hi Benhaynes again, do you know if there is any functional and easy-to-install solution to be able to have a "Repeater" with many fields types, and one of them "Image" (or file)? I have seen the link:

[https://github.com/stouch/directus-interface-uuid-file-image](https://github.com/stouch/directus-interface-uuid-file-image)

and we are trying to take that option, but I don't know if you have any official solution?

thanks!

**Reply by benhaynes** [reply]
Yes, that is what a O2M relationship/interface is.

**Reply by emilianovelasco** [reply]
I am sorry for the basic question, but could you explain me the steps to do it? currently I try first:

1) Add Repeater, pressing "Create Field" green button
2) O2M adding new field, but I don't know what type I should add to have the option to select "Image"

and the second option:

1) Add O2M but I can't choose it either, pressing "Create Field" green button

We want to make a "repeater" that has each item the following: "title => input", "comment => wysiwyg" and "images => Image", it is for a book ranking website/app.

thanks ;)

**Reply by benhaynes** [reply]
You can refer to the Docs/guides:
https://docs.directus.io/configuration/relationships/#one-to-many-o2m

**Reply by emilianovelasco** [reply]
Yes, I know how to create an O2M relationship by creating just one field without repeating it. But I seek to repeat, example (It is for a ranking blog, like "Top 10 books of comedy", it is not necessary to have a collection, it is something simpler and faster to read):

Books (Repeater field type)
-- Book (item added)
---- title (Input field type)
---- thumb (Image field type)

With the relations I could only relate between collections, but I could never add a group of fields (and that this group contains an image, the cover of the book), I will continue reading the documentation, but I think that some repeater or similar is essential.

Thank you very much for the recommendation

**Reply by benhaynes** [reply]
You can have as many fields as you want in the related collection. Apologies, I am not understanding your issue. I'd recommend reaching out to community support: https://directus.chat

**Reply by emilianovelasco** [reply]
I will do that, thank you very much :)

**Reply by enzolupia** [reply]
@emilianovelasco  did you manage to solve your problem? I'm trying to accomplish something very similar.

**Comment by benhaynes** [comment]
Consolidating all other related requests into this Discussion. Please read these other comments for additional context... or copy the important bits into this thread:

* https://github.com/directus/directus/discussions/8834
* https://github.com/directus/directus/discussions/9455
* https://github.com/directus/directus/discussions/4688
* https://github.com/directus/directus/discussions/2736
* https://github.com/directus/directus/discussions/2835

Related relational tickets:

* https://github.com/directus/directus/discussions/3884
* https://github.com/directus/directus/discussions/3962

**Comment by leosin** [comment]
My client requirement is this:
In the product introduction, each row contains 2 contents, a visual editor and an atlas. But the number of rows is undefined.

I am currently thinking of using a repeater, but the image field cannot be uploaded in the repeater

So at present, multiple pictures in the visual editor can only be controlled through the front end.

Is there any other way?

<img width="661" alt="image" src="https://user-images.githubusercontent.com/3676670/211543856-8e4e5a1f-beca-4b08-bc59-e2809f6c1062.png">
displayed on the front end

**Reply by shash7** [reply]
Can confirm, this is big issue. Currently I'm implementing a Directus based solution that's replacing Wordpress ACF fields and I can't upload images via the repeater field.

**Reply by pavelstianko** [reply]
I am using One 2 many to simulate ACF repeater field: 
![image](https://user-images.githubusercontent.com/13132639/223057223-ca7cb695-3ed2-463f-9c98-501678117192.png)

Use only "Add button" without "Select button" and delete item in collection on Delete button click.

**Reply by rijkvanzanten** [reply]
> I am using One 2 many to simulate ACF repeater field: 
> 
> ![image](https://user-images.githubusercontent.com/13132639/223057223-ca7cb695-3ed2-463f-9c98-501678117192.png)
> 
> 
> 
> Use only "Add button" without "Select button" and delete item in collection on Delete button click.

This is the recommended approach. The repeater is great for lightweight flat data (like a list of links), but for anything "bigger" (like ACF style use), you should use the relational route (o2m or m2m for repeating-style, or the m2a if you need a more "dynamic" page builder type affair)

**Comment by stouch** [comment]
FTR https://github.com/stouch/directus-interface-uuid-file-image

**Reply by langy** [reply]
Thanks for this. I had to update the dependencies to the latest in order for it to build, but once built, it works as expected. This functionality should be in the core product.

**Comment by rijkvanzanten** [comment]
Closing this as a duplicate of https://github.com/directus/directus/discussions/3884 & https://github.com/directus/directus/discussions/3962
